runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java

```diff
@@ -0,0 +1,296 @@
+        return new TypeToken<ImmutableCorfuTable<K, V>>() {};
+    }
+
+    public ImmutableCorfuTable() {
+        this.mainMap = HashMap.empty();
+        this.secondaryIndexesWrapper = new SecondaryIndexesWrapper<>();
+    }
+
+    public ImmutableCorfuTable(@Nonnull final Index.Registry<K, V> indices) {
+        this.mainMap = HashMap.empty();
+        this.secondaryIndexesWrapper = new SecondaryIndexesWrapper<>(indices);
+    }
+
+    @Override
+    public ImmutableCorfuTable<K, V> getContext(ICorfuExecutionContext.Context context) {
+        return this;
+    }
+
+    /**
+     *
+     * @param key
+     * @return
+     */
+    public V get(@Nonnull K key) {
+        return mainMap.get(key).getOrNull();
+    }
+
+    /**
+     *
+     * @param key
+     * @param value
+     * @return
+     */
+    public ImmutableCorfuTable<K, V> put(@Nonnull K key, @Nonnull V value) {
+        SecondaryIndexesWrapper<K, V> newSecondaryIndexesWrapper = secondaryIndexesWrapper;
+        if (!secondaryIndexesWrapper.isEmpty()) {
+            V prev = get(key);
+            if (prev != null) {
+                newSecondaryIndexesWrapper = newSecondaryIndexesWrapper.unmapSecondaryIndexes(key, prev);
+            }
+
+            newSecondaryIndexesWrapper = newSecondaryIndexesWrapper.mapSecondaryIndexes(key, value);
+        }
+
+        return new ImmutableCorfuTable<>(mainMap.put(key, value), newSecondaryIndexesWrapper);
+    }
+
+    /**
+     *
+     * @param key
+     * @return
+     */
+    public ImmutableCorfuTable<K, V> remove(@Nonnull K key) {
+        SecondaryIndexesWrapper<K, V> newSecondaryIndexesWrapper = secondaryIndexesWrapper;
+        if (!secondaryIndexesWrapper.isEmpty()) {
+            V prev = get(key);
+            if (prev != null) {
+                newSecondaryIndexesWrapper = newSecondaryIndexesWrapper.unmapSecondaryIndexes(key, prev);
+            }
+        }
+
+        return new ImmutableCorfuTable<>(mainMap.remove(key), newSecondaryIndexesWrapper);
+    }
+
+    /**
+     *
+     * @return
+     */
+    public int size() {
+        return mainMap.size();
+    }
+
+    /**
+     *
+     * @param key
+     * @return
+     */
+    public boolean containsKey(@Nonnull K key) {
+        return mainMap.containsKey(key);
+    }
+
+    /**
+     *
+     * @return
+     */
+    public ImmutableCorfuTable<K, V> clear() {
+        return new ImmutableCorfuTable<>(
+                HashMap.empty(),
+                secondaryIndexesWrapper.clear()
+        );
+    }
+
+    /**
+     *
+     * @return
+     */
+    public java.util.Set<K> keySet() {
+        // TODO: Consider alternative to avoid conversion to Java Set.
+        return mainMap.keySet().toJavaSet();
+    }
+
+    /**
+     *
+     * @return
+     */
+    public Stream<java.util.Map.Entry<K, V>> entryStream() {
+        // TODO: Consider alternative to avoid conversion to Java Map.
+        return mainMap.toJavaMap().entrySet().stream();
+    }
+
```

Comment:
Instead of converting to a map (i.e, `toJavaMap`) you can wrap mainMap.toStream() that converts each entry from Tuple to Map.Entry.

Reply:
Actually, we should use `toJavaParallelStream` instead so that the `scanAndFilter` can benefit from parallelization.